docs/proposals/pod-troubleshooting.md

```diff
@@ -0,0 +1,317 @@
+
+The container will continue to be part of the pod spec, allowing automation to
+inspect the pod, but its resources could be freed by the container runtime. The
+container could be started again manually by performing an update on the pod
+that re-adds the container.
+
+### `kubectl exec` a Container
+
+Once these other features have been implemented, we will add an additional
+option to `kubectl exec` which will cause it to patch a running pod, wait for
+the new container to start and attach to its console. This means it will be able
+to reattach to the troubleshooting session using a standard `kubectl attach`.
+
+```
+% kubectl exec -it -m gcr.io/proj/debug nginx bash
+Reattach to this session using: kubectl attach -it -c debug nginx
+If you don't see a command prompt, try pressing enter.
+root@nginx:/#
+```
+
+#### Alternative: Create New `debug` Command
+
+Rather than extending `kubectl exec` we could create `kubectl debug` to handle
+this use case (and others?). This would be similar to things like `oc debug` and
+could fit into longer term troubleshooting strategy, but it's not technically
+necessary. A new command might be less confusing for someone expecting `kubectl
+exec` to mirror `docker exec` exactly.
+
+## User Stories
+
+### Debugging
+
+Samantha has a service that consists of a statically compiled Go binary running
+in a minimal container that is suddenly having trouble connecting to an internal
+service. Her troubleshooting session might resemble:
+
+```
+% kubectl get pods
+NAME          READY     STATUS    RESTARTS   AGE
+neato-5thn0   1/1       Running   0          1d
+% kubectl exec -it -m gcr.io/neato/debug-image neato-5thn0 bash
+root@debug-image-neato-5thn0:/# cat /etc/resolv.conf
+search default.svc.cluster.local svc.cluster.local cluster.local
+nameserver 10.155.240.10
+options ndots:5
+root@debug-image-neato-5thn0:/# dig @10.155.240.10 neato.svc.cluster.local.
+
+; <<>> DiG 9.9.5-9+deb8u6-Debian <<>> @10.155.240.10 neato.svc.cluster.local.
+; (1 server found)
+;; global options: +cmd
+;; connection timed out; no servers could be reached
+```
+
+This leads Samantha to discover that the cluster's DNS service isn't responding.
+
+### Automation
+
+Abe is a security engineer tasked with running security audits across all of his
+company's running containers. Even though his company has no standard base
+image, he's able to audit all containers using:
+
+```
+% for pod in $(kubectl get -o name pod); do
+    kubectl exec -m gcr.io/neato/security-audit $pod /security-audit.sh
+  done
+```
+
+## Alternatives Considered
+
+### Ephemeral container
+
+An earlier version of this proposal suggested running an ephemeral container in
+the pod namespaces. The container would not be added to the pod spec and would
+exist only as long as the process it ran. This has the advantage of behaving
+similarly to the current kubectl exec, but it is opaque and likely violates
```

Comment:
It also has the advantage of not cluttering up the spec with "dead" containers, right?